File: src/main/scala/ru/tinkoff/gatling/amqp/protocol/AmqpProtocolBuilder.scala
package ru.tinkoff.gatling.amqp.protocol

import com.rabbitmq.client.ConnectionFactory
import ru.tinkoff.gatling.amqp.request.AmqpProtocolMessage

import javax.jms.DeliveryMode

case class AmqpProtocolBuilder(
    requestConnectionFactory: ConnectionFactory,
    replyConnectionFactory: ConnectionFactory,
    deliveryMode: Int = DeliveryMode.NON_PERSISTENT,
    messageMatcher: AmqpMessageMatcher = MessageIdMessageMatcher,
    consumerThreadsCount: Int = 1,
    replyTimeout: Option[Long] = None,
    responseTransformer: Option[AmqpProtocolMessage => AmqpProtocolMessage] = None,
    initActions: AmqpChannelInitActions = Nil
) {

  def usePersistentDeliveryMode: AmqpProtocolBuilder    = copy(deliveryMode = DeliveryMode.PERSISTENT)
  def useNonPersistentDeliveryMode: AmqpProtocolBuilder = copy(deliveryMode = DeliveryMode.NON_PERSISTENT)

  def matchByMessageId: AmqpProtocolBuilder     = messageMatcher(MessageIdMessageMatcher)
  def matchByCorrelationId: AmqpProtocolBuilder = messageMatcher(CorrelationIdMessageMatcher)
  def matchByMessage(extractId: AmqpProtocolMessage => String): AmqpProtocolBuilder =
    messageMatcher(AmqpProtocolMessageMatcher(extractId))

  def responseTransform(ext: AmqpProtocolMessage => AmqpProtocolMessage): AmqpProtocolBuilder =
    copy(responseTransformer = Some(ext))
  private def messageMatcher(matcher: AmqpMessageMatcher) = copy(messageMatcher = matcher)

  def replyTimeout(timeout: Long): AmqpProtocolBuilder            = copy(replyTimeout = Some(timeout))
  def consumerThreadsCount(threadCount: Int): AmqpProtocolBuilder = copy(consumerThreadsCount = threadCount)

  def declare(q: AmqpQueue): AmqpProtocolBuilder    = this.copy(initActions = this.initActions :+ QueueDeclare(q))
  def declare(e: AmqpExchange): AmqpProtocolBuilder = this.copy(initActions = this.initActions :+ ExchangeDeclare(e))

  def bindQueue(q: AmqpQueue, e: AmqpExchange, routingKey: String, args: Map[String, Any] = Map.empty): AmqpProtocolBuilder =
    this.copy(initActions = this.initActions :+ BindQueue(q.name, e.name, routingKey, args))

  def build: AmqpProtocol =
    AmqpProtocol(
      requestConnectionFactory,
      replyConnectionFactory,
      deliveryMode,
      replyTimeout,
      consumerThreadsCount,
      messageMatcher,
      responseTransformer,
      initActions
    )
}

File: src/main/java/ru/tinkoff/gatling/javaapi/protocol/AmqpProtocolBuilder.java
package ru.tinkoff.gatling.javaapi.protocol;

import io.gatling.core.protocol.Protocol;
import io.gatling.javaapi.core.ProtocolBuilder;
import ru.tinkoff.gatling.amqp.protocol.AmqpProtocol;
import ru.tinkoff.gatling.amqp.request.AmqpProtocolMessage;
import static scala.jdk.javaapi.CollectionConverters.asScala;
import scala.Function1;

import java.util.Map;

public class AmqpProtocolBuilder implements ProtocolBuilder{
    private ru.tinkoff.gatling.amqp.protocol.AmqpProtocolBuilder wrapped;

    public AmqpProtocolBuilder(ru.tinkoff.gatling.amqp.protocol.AmqpProtocolBuilder wrapped) {
        this.wrapped = wrapped;
    }

    public AmqpProtocolBuilder usePersistentDeliveryMode() {
        this.wrapped = wrapped.usePersistentDeliveryMode();
        return this;
    }

    public AmqpProtocolBuilder useNonPersistentDeliveryMode() {
        this.wrapped = wrapped.useNonPersistentDeliveryMode();
        return this;
    }

    public AmqpProtocolBuilder matchByMessageId() {
        this.wrapped = wrapped.matchByMessageId();
        return this;
    }

    public AmqpProtocolBuilder matchByCorrelationId() {
        this.wrapped = wrapped.matchByMessageId();
        return this;
    }

    public AmqpProtocolBuilder matchByMessage(Function1<AmqpProtocolMessage, String> extractId) {
        this.wrapped = wrapped.matchByMessage(extractId);
        return this;
    }

    public AmqpProtocolBuilder responseTransform(Function1<AmqpProtocolMessage, AmqpProtocolMessage> ext) {
        this.wrapped = wrapped.responseTransform(ext);
        return this;
    }

    public AmqpProtocolBuilder replyTimeout(Long timeout) {
        this.wrapped = wrapped.replyTimeout(timeout);
        return this;
    }

    public AmqpProtocolBuilder consumerThreadsCount(Integer threadCount) {
        this.wrapped = wrapped.consumerThreadsCount(threadCount);
        return this;
    }

    public AmqpProtocolBuilder declare(AmqpQueue q) {

        this.wrapped = wrapped.declare(ru.tinkoff.gatling.amqp.Predef.queue(
                q.getName(),
                q.getDurable(),
                q.getExclusive(),
                q.getAutoDelete(),
                scala.collection.immutable.Map.from(asScala(q.getArguments()))
        ));
        return this;
    }

    public AmqpProtocolBuilder declare(AmqpExchange e) {
        this.wrapped = wrapped.declare(ru.tinkoff.gatling.amqp.Predef.exchange(
                e.getName(),
                e.getExchangeType(),
                e.getDurable(),
                e.getAutoDelete(),
                scala.collection.immutable.Map.from(asScala(e.getArguments()))
        ));
        return this;
    }

    public AmqpProtocolBuilder bindQueue(
            AmqpQueue q,
            AmqpExchange e,
            String routingKey,
            Map<String, Object> args
    ) {
        this.wrapped = wrapped.bindQueue(
                ru.tinkoff.gatling.amqp.Predef.queue(
                        q.getName(),
                        q.getDurable(),
                        q.getExclusive(),
                        q.getAutoDelete(),
                        scala.collection.immutable.Map.from(asScala(q.getArguments()))
                ),
                ru.tinkoff.gatling.amqp.Predef.exchange(
                        e.getName(),
                        e.getExchangeType(),
                        e.getDurable(),
                        e.getAutoDelete(),
                        scala.collection.immutable.Map.from(asScala(e.getArguments()))
                ),
                routingKey,
                scala.collection.immutable.Map.from(asScala(args))
        );
        return this;
    }

    @Override
    public Protocol protocol() {
        return wrapped.build();
    }
}
